src/main/java/com/google/devtools/build/lib/remote/AbstractActionInputPrefetcher.java

```diff
@@ -183,13 +182,16 @@ protected ListenableFuture<Void> prefetchFiles(
     Map<SpecialArtifact, List<TreeFileArtifact>> trees = new HashMap<>();
     List<ActionInput> files = new ArrayList<>();
     for (ActionInput input : inputs) {
-      checkArgument(!input.isDirectory(), "cannot prefetch a directory");
-
       // Source artifacts don't need to be fetched.
       if (input instanceof Artifact && ((Artifact) input).isSourceArtifact()) {
         continue;
       }
 
+      // Skip empty tree artifacts (non-empty tree artifacts should have already been expanded).
+      if (input.isDirectory()) {
+        continue;
```

Comment:
Do we want to check whether the directory is actually empty?

Reply:
Unfortunately, the `FileArtifactValue` returned by `MetadataHandler#getMetadata` for a tree artifact doesn't provide enough information to do the check, and I'd rather not add API surface just for it.

I wonder if we could do this in a more principled way, by having the ArtifactExpander produce a special ActionInput that signals "this is root directory for an empty tree artifact"? But that's clearly for a separate PR.

Reply:
I started working on that in https://github.com/bazelbuild/bazel/pull/16015/files. If you are also interested in this, we could continue the discussion there.